### PR TITLE
[3.6] More improvements to New Storage Layer Compatibility

### PR DESCRIPTION
**Pull request by rossriley**
This is targeted at installs where legacy `setcontent` is turned off so the Content objects are filled using the new storage layer.

As discussed in #6985 methods on the content object itself are moved to twig functions and filters with a compatibility patch used to avoid things breaking until 4.0.

There's also some ongoing work on performance here, targeted at closing some performance problems with content object hydration.

Existing systems using the legacy setcontent system will continue to work as before.

### CODE REVIEW

**Comment by rossriley** [comment]
OK, good to go.

For testing this remember you need to make the below addition to config.yml

```
compatibility:
    setcontent_legacy: false
```

**Comment by bobdenotter** [comment]
I've noticed that "previewing" with Blocks / Repeaters is broken: 

screen shot 2018-01-22 at 20 59 56

This is something I can(should) fix in the templates themselves, though. Will do that after this is merged in. 

Other than that, i didn't bump into any quirks yet, so that's great. I'll update a non-trivial site tomorrow, and see if it works smoothly, too. 👍

**Review by bobdenotter — DISMISSED** [review]

**Comment by rossriley** [comment]
@bob I've had a bit of an attempt at fixing this in the storage layer but because we need to get it to work in both legacy and new storage I'm thinking that you were probably correct in suggesting fixing this at the template level.

We have a method in the `storage.metadata` service `getFieldTypeFor($contentType, $column, $field = null)` which pulls back the field type info for any ct/field combination and optionally with the third parameter for the sub-field of a repeater.

What I would suggest is that we expose this method via a Twig function and then rather than calling: `{% set fieldtype = repeaterfield.fieldtype() %}` as we currently do in the _sub_fields template (which fails when repeaterfield is an array) we call the new function using the contenttype/field/repeatername and avoid this error.

I can either add this to this PR or we can handle separately, whatever works best.

**Comment by bobdenotter** [comment]
> What I would suggest is that we expose this method via a Twig function

Sounds good to me. If you add it to this PR, i'll update the themes accordingly.

**Comment by rossriley** [comment]
ok, the new filter/function is included here and I've tested with the base-2016 theme and all seems to be ok. Here's what needs to be modified...

Firstly in the outer loop: `{% for key, value in record.values if (key not in omittedkeys) %}`
any call to `record.fieldtype(key)` simpley gets changed to `record|fieldtype(key)`

Secondly in the inner loop for repeaters and blocks we need to keep access to the original key variable so one minor change to the for loop: `{% for repeatkey, repeaterfield in repeater %}` that is using repeatkey instead of key.

Then final changes, to get the fieldtype now in this repeater loop: 
`{% set fieldtype = record|fieldtype(key, repeatkey) %}` same as above except we also pass in the sub-field name.

Finally to get the value we need to check whether we have an object or array so the final block looks like:
```
                {% for repeater in value %}
                    {% for repeatkey, repeaterfield in repeater %}

                        {% set fieldtype = record|fieldtype(key, repeatkey) %}
                        {% if attribute(repeaterfield, 'getValue') is defined %}
                            {% set value = repeaterfield.value() %}
                        {% else %}
                            {% set value = repeaterfield %}
                        {% endif %}
                        {{ block('common_fields') }}
                        {{ block('extended_fields') }}

                    {% endfor %}
                {% endfor %}
```

**Comment by JarJak** [comment]
I get:
```
Exception in Storage.php line 78:
\MyExtension\CustomContentClass does not extend \Bolt\Legacy\Content.
```
Looks like custom content classes still need to extend legacy :(

It happens because new EntityManager uses old Storage, which does callback to itself in line 1627 (callback is defined in line 1055).

**Comment by JarJak** [comment]
And if I extend \Bolt\Legacy\Content then I get:
`An exception has been thrown during the rendering of a template ("Notice: Undefined variable: pager")`
in line:
```
{% setcontent url = ct.slug where {ID: record.relation.urls|first} returnsingle %}
```

**Comment by rossriley** [comment]
@JarJak Once this is merged into 3.6 I'll revisit, there's been some changes to the default theme since and I'm guessing a few things have broken since.

**Comment by JarJak** [comment]
@rossriley I don't think this can be merged, it is still buggy. And I don't think these bugs have anything in common with theme updates.

**Comment by GwendolenLynch** [comment]
> I don't think this can be merged,

Yeah, we're still in review phase and have talked about this PR at pretty much every dev meeting this year :laughing:

**Comment by JarJak** [comment]
I know :D But I wanted to say that it needs to be revisited before merge,
not after.

File: src/Storage/Query/Query.php
Comment (JarJak) [inline src/Storage/Query/Query.php:98]: why view and model layers are mixed here?

File: src/Twig/Extension/ArrayExtension.php
Comment (JarJak) [inline src/Twig/Extension/ArrayExtension.php:65]: silencing invalid argument? what's the purpose of this?

File: src/Controller/Base.php
Comment (JarJak) [inline src/Controller/Base.php:427]: 1. Is this a BC Break?
2. Docs and config samples should be updated too.
3. I'm 99% sure that this param is also used somewhere else.

**Review by JarJak — CHANGES_REQUESTED** [review]

File: src/Controller/Base.php
Comment (GwendolenLynch) [reply src/Controller/Base.php:427]: > 3. I'm 99% sure that this param is also used somewhere else.

```
$ ag log_not_found
src/Controller/Frontend.php
152:        $content = $this->getContent($contenttype['slug'], ['slug' => $slug, 'returnsingle' => true, 'log_not_found' => !is_numeric($slug)]);

src/Legacy/Storage.php
1602:        if (isset($parameters['log_not_found'])) {
1603:            $logNotFound = $parameters['log_not_found'];
1604:            unset($parameters['log_not_found']);
```

In the case of the controller, that is only going to legacy, and legacy is legacy.

Comment (rossriley) [reply src/Controller/Base.php:427]: Yes, this is only a legacy feature and I don't think the query parser is the right place for it. We have HTTP status exceptions and they can look after logging 404 errors if so needed. As with everything in this PR legacy stays exactly as is and we can decide for 4.0 onwards what we want to keep and drop. If in this testing phase a feature is found to be missing, then we will happily add it in, but there's no need to automatically include existing features if we think they don't fit.

Comment (JarJak) [reply src/Controller/Base.php:427]: Oh ok, now I can see that this is unset only for new Storage, legacy one still receives this parameter. 👍

**Review by JarJak — APPROVED** [review]

**Review by JarJak — CHANGES_REQUESTED** [review]
just reverting approval

**Review by GwendolenLynch — APPROVED** [review]
Let's get this in and under test :+1:

**Comment by JarJak** [comment]
@GawainLynch I see this got merged without changes that I've requested and without any explanation, why is that?

**Comment by GwendolenLynch** [comment]
Your opinions/reviews are **very much** welcomed, and considered, but you are not a core team member, so not in possession of a blocking vote. :wink: 

Further, the weekly meetings are a much better place to go over finer details. While I agree with some of your review comments, they also partially miss some of what needs to be balanced out to get this across the line :smile: 

That said, PRs to RFC additional changes are also welcome. But the core team members are heavily stretched in RL this year, so "keeping things moving" is important.

**Comment by JarJak** [comment]
Thanks for explanation. I'm fine with that I'm not a decider ;) just wanted to know why it got merged without replies.

I will test it in real project and report issues if any of them will pop out :)

**Comment by GwendolenLynch** [comment]
> I will test it in real project and report issues if any of them will pop out :)

Please! All help & bug reports/PRs appreciated.

**Comment by bobdenotter** [comment]
> Further, the weekly meetings are a much better place to go over finer details. While I agree with some of your review comments, they also partially miss some of what needs to be balanced out to get this across the line 😄

If you want to, you're free to pop in tomorrow (tuesdays, 19:30 CET), to discuss additional changes with us! :+1:

**Pull request merged**
